manager/dispatcher/dispatcher.go

```diff
@@ -759,6 +759,7 @@ func (d *Dispatcher) Assignments(r *api.AssignmentsRequest, stream api.Dispatche
 		initial   api.AssignmentsMessage
 	)
 	tasksMap := make(map[string]*api.Task)
+	tasksUsingSecret := make(map[string]map[string]struct{})
```

Comment:
Not your fault, but this function was already way too long and way too complex and this PR makes it even more so.

We're topping nearly 400 lines of complex code with anonymous functions and gotos, this function needs to be refactored/rethought.

This is one of the most critical functions in our code base. It started at perhaps 30 lines and gradually got out of control.

/cc @diogomonica @dongluochen @aaronlehmann